File: Cadena de Hoteles/Hotel/src/main/java/org/example/repositories/LogReservacionRepository.java
package org.example.repositories;

import org.example.data.DatabaseManager;

/**
 * Repository para el registro de eventos de reservacion en la tabla LogReservacion.
 */
public class LogReservacionRepository {

    public static final int TIPO_RESERVACION_EXITOSA         = 1;
    public static final int TIPO_RESERVACION_FALLIDA         = 2;
    public static final int TIPO_RESERVACION_ERROR_INTERNO   = 3;
    public static final int TIPO_AGENCIA_EXITOSA             = 4;
    public static final int TIPO_AGENCIA_FALLIDA             = 5;
    public static final int TIPO_AGENCIA_ERROR               = 6;
    public static final int TIPO_AGENCIA_EXPIRADA            = 7;

    public static final int TIPO_PAGO_EXITOSO         = 8;
    public static final int TIPO_PAGO_FALLIDO         = 9;
    public static final int TIPO_PAGO_ERROR_INTERNO   = 10;
    public static final int TIPO_PAGO_AGENCIA_EXITOSO = 11;
    public static final int TIPO_PAGO_AGENCIA_FALLIDO = 12;
    public static final int TIPO_PAGO_AGENCIA_ERROR   = 13;

    public static final int TIPO_CANCELACION_EXITOSA         = 14;
    public static final int TIPO_CANCELACION_FALLIDA         = 15;
    public static final int TIPO_CANCELACION_AGENCIA_EXITOSA = 16;
    public static final int TIPO_CANCELACION_AGENCIA_FALLIDA = 17;

    public static final int TIPO_RESERVACION_EXPIRADA_AUTO = 18;

    /**
     * Registra un evento de reservacion en la tabla LogReservacion.
     * Si el INSERT falla, imprime el error en consola sin interrumpir el flujo principal.
     *
     * @param tipoEventoId  ID del tipo de evento (usar constantes TIPO_* de esta clase).
     * @param reservacionId ID de la reservacion creada, null si la creacion fallo.
     * @param usuarioId     ID del usuario autenticado, null si es reservacion de agencia.
     * @param agenciaId     ID de la agencia, null si es reservacion de usuario comun.
     * @param noReservacion codigo MIKU-XXXXXXXX, null si la creacion fallo.
     * @param total         total calculado de la reservacion, null si fallo antes del calculo.
     * @param exitoso       true si el evento fue exitoso.
     * @param ip            IP del cliente.
     * @param userAgent     User-Agent del cliente.
     * @param mensaje       detalle adicional del evento, puede ser null.
     */
    public void registrar(int tipoEventoId, Integer reservacionId, Integer usuarioId,
                          Integer agenciaId, String noReservacion, Double total,
                          boolean exitoso, String ip, String userAgent, String mensaje) {
        try {
            DatabaseManager.executeUpdate(
                    "INSERT INTO LogReservacion " +
                            "(TipoEventoID, ReservacionID, UsuarioID, AgenciaID, NoReservacion, " +
                            " TotalCalculado, Exitoso, IPOrigen, UserAgent, Mensaje, Fecha) " +
                            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)",
                    tipoEventoId,
                    reservacionId,
                    usuarioId,
                    agenciaId,
                    noReservacion,
                    total,
                    exitoso ? 1 : 0,
                    ip,
                    userAgent,
                    mensaje
            );
        } catch (Exception e) {
            System.err.println("[LogReservacionRepository] Error al registrar log (tipo="
                    + tipoEventoId + "): " + e.getMessage());
        }
    }
}

File: Aero Linea/Aerolinea.API/Aerolinea.API/Repositories/LogReservacionRepository.cs
using Aerolinea.API.Data;
using Microsoft.Data.SqlClient;

namespace Aerolinea.API.Repositories
{
    /// <summary>
    /// Repositorio para el registro de eventos de reservacion en la tabla LogReservacion.
    /// Si el INSERT falla, imprime el error en consola sin interrumpir el flujo principal.
    /// </summary>
    public class LogReservacionRepository
    {
        public const int TipoReservacionExitosa = 1;
        public const int TipoReservacionFallida = 2;
        public const int TipoReservacionErrorInterno = 3;
        public const int TipoAgenciaExitosa = 4;
        public const int TipoAgenciaFallida = 5;
        public const int TipoAgenciaError = 6;
        public const int TipoAgenciaExpirada = 7;
        public const int TipoPasajerosAgregados = 8;
        public const int TipoPasajerosError = 9;
        public const int TipoPasajerosAgenciaAgregados = 10;
        public const int TipoPasajerosAgenciaError = 11;

        public const int TipoPagoExitoso = 12;
        public const int TipoPagoFallido = 13;
        public const int TipoPagoErrorInterno = 14;
        public const int TipoPagoAgenciaExitoso = 15;
        public const int TipoPagoAgenciaFallido = 16;
        public const int TipoPagoAgenciaError = 17;

        public const int TipoCancelacionExitosa = 18;
        public const int TipoCancelacionFallida = 19;
        public const int TipoCancelacionAgenciaExitosa = 20;
        public const int TipoCancelacionAgenciaFallida = 21;

        public const int TipoReservacionExpiradaAuto = 22;

        private readonly DbConnectionFactory _connectionFactory;

        public LogReservacionRepository(DbConnectionFactory connectionFactory)
        {
            _connectionFactory = connectionFactory;
        }

        /// <summary>
        /// Registra un evento de reservacion en la tabla LogReservacion.
        /// </summary>
        public async Task Registrar(int tipoEventoId, int? reservacionId, int? usuarioId,
                                    int? agenciaId, decimal? total, bool exitoso,
                                    string? ip, string? userAgent, string? mensaje)
        {
            try
            {
                using var connection = _connectionFactory.CreateConnection();
                await connection.OpenAsync();

                var sql = @"
                    INSERT INTO LogReservacion
                        (TipoEventoID, ReservacionID, UsuarioID, AgenciaID,
                         TotalCalculado, Exitoso, IPOrigen, UserAgent, Mensaje, Fecha)
                    VALUES
                        (@TipoEventoID, @ReservacionID, @UsuarioID, @AgenciaID,
                         @TotalCalculado, @Exitoso, @IPOrigen, @UserAgent, @Mensaje, GETDATE())";

                using var command = new SqlCommand(sql, connection);
                command.Parameters.AddWithValue("@TipoEventoID", tipoEventoId);
                command.Parameters.AddWithValue("@ReservacionID", (object?)reservacionId ?? DBNull.Value);
                command.Parameters.AddWithValue("@UsuarioID", (object?)usuarioId ?? DBNull.Value);
                command.Parameters.AddWithValue("@AgenciaID", (object?)agenciaId ?? DBNull.Value);
                command.Parameters.AddWithValue("@TotalCalculado", (object?)total ?? DBNull.Value);
                command.Parameters.AddWithValue("@Exitoso", exitoso);
                command.Parameters.AddWithValue("@IPOrigen", (object?)ip ?? DBNull.Value);
                command.Parameters.AddWithValue("@UserAgent", (object?)userAgent ?? DBNull.Value);
                command.Parameters.AddWithValue("@Mensaje", (object?)mensaje ?? DBNull.Value);

                await command.ExecuteNonQueryAsync();
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"[LogReservacionRepository] Error al registrar log (tipo={tipoEventoId}): {e.Message}");
            }
        }
    }
}